### PR TITLE
Initial take at adding JMH micro-benchmarks

### LINKED ISSUE #12641: ability to run JMH benchmarks from gradle
### Description

Background: I'm having a hard time keeping https://github.com/rmuir/vectorbench up to date, the code has differences with what the integrated vector code in lucene is, I have to copy/paste back and forth to run tests vs microbenchmarks, it is quite tedious and error-prone.

I think it would be better to just have benchmark files that we can run from here? I tried my hand at https://github.com/melix/jmh-gradle-plugin but it didn't seem to play well with things like `apply plugin` conditionally, or our alt-jvm usage. Probably easier to just have a simple JavaExec-type task that launches it the way we want?

IMO: it would be great to just be able to run `./gradlew microbench` or something to run these benchmarks. And to call our actual code, not a copy-paste of it (e.g. VectorUtil.xxxx).

### PR DESCRIPTION
**Pull request by dweiss**
As per discussion in: https://github.com/apache/lucene/issues/12641

This patch adds the ability to compile JMH microbenchmarks and run them from within Lucene codebase. I didn't use any plugins as they, in my opinion, obscure the view of what's happening and what's needed. I also wanted it to work with the module system and without uber-jars.

```
gradlew -p lucene/benchmark-jmh assemble
```

compiles microbenchmarks with JMH, then displays a short help on what to do next:
```
JMH benchmarks compiled. Run them with:

java -jar lucene\benchmark-jmh\build\benchmarks\lucene-benchmark-jmh-10.0.0-SNAPSHOT.jar

or

java --module-path lucene\benchmark-jmh\build\benchmarks --module org.apache.lucene.benchmark.jmh

JMH options you can use with the above:

  -h      displays verbose help for all options
  -l      list available benchmarks
  -lp     list benchmarks that pass the filter and their parameters
  regexp  execute all benchmark containing regexp
```

I didn't add any gradle-specific tasks to run the benchmarks directly - it's technically easy but it seems like a bad idea to fork a subprocess from (quite heavy) gradle when such delicate things are being measured? I'm not sure here.

Some future improvements/ remaining issues include:

1) multi-source compilation so that code can be compiled against future APIs (not the one gradle runs with).
2) forbiddenapis doesn't have access to vectors and I had to turn it off,
3) jmh is GPL. It's a dev tool, I don't think it's a problem. I left license checks out of it for the time being.

### CODE REVIEW

**Comment by rmuir** [comment]
@dweiss thank you for this! I tried it out and it works great. Glad to see the `cosineDistanceVectorUtil()` working in this example which is exactly what I want to do.

**Comment by dweiss** [comment]
Let me review the minor remaining bits - I had to leave the computer before the tests completed.

**Comment by benwtrent** [comment]
@dweiss this is awesome!!! Thank you so much!

**Comment by benwtrent** [comment]
@dweiss one suggestion I have is adding a `README` for `lucene/benchmark-jmh` on how to run the benchmarks and add new ones. Your description on how to use it in this PR seems adequate.

I tested this locally and was able to follow your directions and get the benchmarks to run without issue.

**Comment by dweiss** [comment]
I enabled forbiddenapis to the extent possible, to make Uwe happier. ;) I like those flame-graphs that Solr has, etc., but these can come later. I'll add a simple ```gradlew :helpJmh``` target similar to everything else.

**Comment by dweiss** [comment]
@benwtrent Can you take a look? If there's any wording you think would work better, please do change it. Explaining how JMH works is perhaps beyond the scope of the readme file... ;)

File: lucene/benchmark-jmh/README.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+command line, for example:
+
+java --module-path lucene\benchmark-jmh\build\benchmarks --module org.apache.lucene.benchmark.jmh
+
+You can pass any JMH options to the above command, for example:
+
+  -h      displays verbose help for all options
+  -l      list available benchmarks
+  -lp     list benchmarks that pass the filter and their parameters
+  regexp  execute all benchmark containing regexp
```
Comment (benwtrent) [inline lucene/benchmark-jmh/README.txt:21]:
```suggestion
  regexp  execute all benchmark containing regexp
 
Here is an example running a single benchmark:

java --module-path lucene\benchmark-jmh\build\benchmarks --module org.apache.lucene.benchmark.jmh org.apache.lucene.benchmark.jmh.BinaryCosineBenchmark.cosineDistanceNew

Or running all of BinaryCosineBenchmark 

java --module-path lucene\benchmark-jmh\build\benchmarks --module org.apache.lucene.benchmark.jmh org.apache.lucene.benchmark.jmh.BinaryCosineBenchmark
```

**Comment by benwtrent** [comment]
Looks good @dweiss I added a line with an example for running a single benchmark or suite of benchmarks.

File: lucene/benchmark-jmh/README.txt (hunk 1 above)
Comment (dweiss) [reply lucene/benchmark-jmh/README.txt:21]: Oh, apologies - I thought you added that change to the branch, not as github's suggested change. I added this manually.

**Pull request merged**